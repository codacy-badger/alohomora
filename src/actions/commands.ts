export { command as listCommand } from './listParameters';
export { command as getCommand } from './getParameter';