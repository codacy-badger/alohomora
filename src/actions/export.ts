// type ExportOptions = { template?: string = 'shell', custom?: string };

// const exportOutput = (options: Options, template?: string = 'shell', custom?: string): string => {

// }
