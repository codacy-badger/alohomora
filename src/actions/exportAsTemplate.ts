import SSM from 'aws-sdk/clients/ssm';
import groupBy from 'lodash.groupby';

import { Options } from "../types";
import { REGION, API_VERSION, MAX_RESULTS_FOR_PATH, Environment } from '../utils/constants';
import { normalizeSecretKey } from '../utils/normalizeSecretKey';
interface Input extends Options { templateName?: string, custom?: string };

const getParametersByPath = async (params: SSM.GetParametersByPathRequest, region: string): Promise<SSM.ParameterList> => {

  const ssm = new SSM({ apiVersion: API_VERSION, region });
  const { Parameters: parameters = [], NextToken } = await ssm.getParametersByPath(params).promise();

  if (NextToken) {
    const moreParameters = await getParametersByPath({ ...params, NextToken }, region);
    return [...parameters, ...moreParameters];
  } else {
    return parameters;
  }
}
const getKeys = async ({ prefix, region }: { prefix: string, region: string }) => {
  const parameters = await getParametersByPath({
    Path: `/${prefix}`,
    MaxResults: MAX_RESULTS_FOR_PATH,
    Recursive: true,
    WithDecryption: true,
  }, region);

  const secretPairs = parameters.map(({ Name = '', Value: value = '' }) => {

    const { name, environment } = normalizeSecretKey(Name, prefix);
    return {
      name,
      environment,
      value,
    }
  });
  return groupBy(secretPairs, ({ environment }: { environment: string }) => environment);
}
export const exportAsTemplate = async ({ prefix, environment = Environment.all, region = REGION }: Input): Promise<string> => {

  try {
    const groupedSecrets = await getKeys({
      prefix,
      region,
    });

    const baseSecrets = groupedSecrets[Environment.all].reduce(((memo, secret) => {
      return Object.keys(memo).length ? { ...memo, [secret.name]: secret.value } : { [secret.name]: secret.value };
    }), {} as Record<string, string>);;

    if (environment !== Environment.all && environment in groupedSecrets) {
      const envSecrets: Record<string, string> = groupedSecrets[environment].reduce(((memo, secret) => {
        return { ...memo, [secret.name]: secret.value };
      }));

      return JSON.stringify({ ...baseSecrets, ...envSecrets }, null, 2);
    } else {
      return JSON.stringify(baseSecrets, null, 2);
    }
  } catch (e) {
    console.error('something went awfully wrong', e);
    throw e;
  }
};
