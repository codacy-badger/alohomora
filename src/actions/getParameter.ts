import SSM from 'aws-sdk/clients/ssm';
import dateFormat from 'dateformat'
import Table from 'cli-table3';
import ora from 'ora';

import { Options } from '../types';
import { REGION, API_VERSION, Environment, DATE_FORMAT, SUCCESS_SYMBOL } from '../utils/constants';


interface Input extends Options {
  name: string
};

export const getParameter = async ({ name, prefix, region = REGION, environment = Environment.all }: Input): Promise<string> => {

  const loader = ora(`retrieving key ${name} with the prefix /${prefix}  (${region})`).start();

  const ssm = new SSM({ apiVersion: API_VERSION, region });
  const params = {
    Name: `/${prefix}/${environment}/${name}`,
    WithDecryption: true
  };

  const table = new Table({
    head: ['Name', 'Value', 'Environment', 'Updated by', 'Version']
  });


  let response;

  try {
    response = await ssm.getParameter(params).promise();

    if (response.Parameter) {

      const { Value: value, LastModifiedDate: updatedAt, Version: version } = response.Parameter;

      table.push([name, value, environment, dateFormat(updatedAt, DATE_FORMAT), version]);

      loader.stopAndPersist({ text: `${name} found under /${prefix}  (${region})`, symbol: SUCCESS_SYMBOL });
      return table.toString();

    } else {
      loader.stopAndPersist({ text: `${name} not found under /${prefix}  (${region})`, symbol: SUCCESS_SYMBOL });
    }

  } catch (e) {
    loader.fail(`something went wrong retrieving the key ${name}, ${e}`);
  }

  return '';
}