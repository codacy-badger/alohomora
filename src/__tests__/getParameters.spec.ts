import SSM from '../__mocks__/aws-sdk/clients/ssm';
import { getParameter } from '../actions';

const getParameterPayload = {
  "Name": "/my-company/my-app/production/Vault_713",
  "Type": "String",
  "Value": "Philosopher's Stone",
  "Version": 1,
  "LastModifiedDate": "2020-04-26T02:03:31.132Z",
  "ARN": "arn:aws:ssm:us-west-2:687711712713:parameter/my-company/my-app/production/Vault_713",
  "DataType": "text"
};

describe("getParameters", () => {
  it("request fails", async () => {
    const prefix = 'my-company/my-app';

    const handler = jest.fn()
      .mockImplementationOnce(() => { throw new Error('Some fatal error ocurred') });

    SSM.__setResponseForMethods({ getParameter: handler });

    const response = await getParameter({ name: 'Vault_713', environment: 'production', prefix });
    expect(response).toBe('');

  });

  it("gets the parameter", async () => {
    const prefix = 'my-company/my-app';

    const handler = jest.fn(() => ({ Parameter: getParameterPayload }));

    SSM.__setResponseForMethods({ getParameter: handler });

    const response = await getParameter({ name: 'Vault_713', environment: 'production', prefix });
    expect(response).toMatchSnapshot();
  });

  it("does not find the parameter", async () => {
    const prefix = 'my-company/my-app';

    const error = { code: 'ParameterNotFound', name: 'ParameterNotFound' };

    const handler = jest.fn(() => { throw error; });

    SSM.__setResponseForMethods({ getParameter: handler });

    const response = await getParameter({ name: 'Vault_713', environment: 'production', prefix });
    expect(response).toBe('');
  })

  it("does not find the parameter, but it returns Parameter empty, without throwing an Error", async () => {
    const prefix = 'my-company/my-app';

    const handler = jest.fn(() => ({ Parameter: undefined }));

    SSM.__setResponseForMethods({ getParameter: handler });

    const response = await getParameter({ name: 'Vault_713', environment: 'production', prefix });
    expect(response).toBe('');
  })
});
