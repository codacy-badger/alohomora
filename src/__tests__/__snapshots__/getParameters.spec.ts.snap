// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getParameters gets the parameter 1`] = `
"[90m┌───────────[39m[90m┬─────────────────────[39m[90m┬─────────────[39m[90m┬──────────────────────────────────────────────[39m[90m┬─────────┐[39m
[90m│[39m[31m Name      [39m[90m│[39m[31m Value               [39m[90m│[39m[31m Environment [39m[90m│[39m[31m Updated by                                   [39m[90m│[39m[31m Version [39m[90m│[39m
[90m├───────────[39m[90m┼─────────────────────[39m[90m┼─────────────[39m[90m┼──────────────────────────────────────────────[39m[90m┼─────────┤[39m
[90m│[39m Vault_713 [90m│[39m Philosopher's Stone [90m│[39m production  [90m│[39m Saturday, April 25th, 2020, 7:03:31 PM (PDT) [90m│[39m 1       [90m│[39m
[90m└───────────[39m[90m┴─────────────────────[39m[90m┴─────────────[39m[90m┴──────────────────────────────────────────────[39m[90m┴─────────┘[39m"
`;
